### DISCUSSION TITLE
Tasmota smart meter interface Landis+Gyr T550 MQTT record always 0

### DISCUSSION THREAD

**Discussion by jjeb-de** [arendst/Tasmota#20715] [Q&A]
Hi everybody,

I am stuck with the following problem and appreciate some help.

I have a smart meter Landis+Gyr T550 with an optical interface. My reading device is bitshake smart meter reader air with tasmota version 13.
I created a script based on a template, see following
[script.txt](https://github.com/arendst/Tasmota/files/14258469/script.txt)

dumping the received data looks quite ok, see following
[result.txt](https://github.com/arendst/Tasmota/files/14258484/result.txt)

But the MQTT record sent to my MQTT broker contains all zeros, see following
[mqtt.txt](https://github.com/arendst/Tasmota/files/14258520/mqtt.txt)

Any idea, what could be the reason for this behaviour?

Thanks in advance.

**Comment by gemu2015** [comment]
it seems that this meter does not emit a CR after each value. thus it can not be used in "line mode" which is default 
try switching to shift mode by adding this line to the descriptor
1,=so2,0

**Comment by jjeb-de** [comment]
Hi gemu2015,

Thanks for your fast feedback. I just changed the descriptor setting according your proposal, now I understand the chapter special commands in the smart meter interface description.

**Comment by jjeb-de** [comment]
Hi gemu2015,

sorry for the late response, works perfect, thanks

**Reply by Mart-Muc** [reply]
hi JJeb-de

Kannst du bitte das ganze und funktionierende Script uploaden- ich kann leider mit [gemu2015](https://github.com/gemu2015)s Anweisung nix anfangen, sorry.

Danke

**Reply by jjeb-de** [reply]
hier das script wie es bei mir nun läuft:

-------------------------------------------------------------------------------------------------------------------------------------
>D
scnt=0
res=0

>B
;reset device interface
->sensor53 r
;disable publishing at MQTT teleperiod, on boot
smlj=0

>S
;re-enable publishing at MQTT teleperiod, on 10 seconds of up-time
if upsecs>10
then
smlj|=1
endif

>F
; 100ms iteration
scnt+=1

;actions
switch scnt
;start after 1.6 seconds (16 x 100ms)
case 16
;set transfer speed to 300 baud
res=sml(1 0 300)
;trigger meter by sending 40 times 'NUL' followed by '/?! CR LF'
res=sml(1 1 "0000000000000000000000000000000000000000")
res=sml(1 1 "0000000000000000000000000000000000000000")
res=sml(1 1 "2F3F210D0A")
case 32
;request transfer speed by sending 'ACK 0 x 0 CR LF'
; x hex baud
; '0' = 30 = 300
; '1' = 31 = 600
; '2' = 32 = 1200
; '3' = 33 = 2400
; '4' = 34 = 4800
; '5' = 35 = 9600
; '6' = 36 = 19200
; '7' = 37 = 38400
; '8' = 38 = 76800
; '9' = 39 = 115200
;transfer speed 2400 baud
res=sml(1 1 "063033300D0A")
res=sml(1 0 2400)
;restart sequence after 10 minutes (6000 x 100 ms)
case 6000
scnt=0
ends

;parse and store
; Device: Landis+Gyr T550 (UH50)
; Protocol: OBIS Ascii line shift mode
; Serial setting: 2400@7E1 for OBIS Ascii,
>M 1
+1,5,o,0,2400,LGUH50,4
1,=so2,0
1,6.8(@1,Zählerstand,MWh,mwh_count,3
1,6.6(@1,Max Heizleistung,kWh,kwh_max_heatperf,1
1,6.26(@1,Durchlauf,m³,m3_throughput,2
1,6.33(@1,Max Durchlauf per h,m³h,m3h_max_throughput,3
;1,9.4(@#),Max Temp VR-Lauf,°C,c_max_temp_frw,0
1,9.4(@1,Max Temp V-Lauf,°C,c_max_temp_fw,1
1,9.4(@s&1:1,Max Temp R-Lauf,°C,c_max_temp_rw,1

-------------------------------------------------------------------------------------------------------------------------------------

**Comment by l33** [comment]
Once shift mode is enabled, how can OBIS codes which are a substring of others be correctly parsed? For example what if there is 6.8 and 3.6.8 and the longer code appears after the shorter one in the data? In those cases where the record is in the middle of the line it is typically preceded by a `)`. Adding the bracket to the definition line such as `1,)6.8(` parses the data correctly. However, I was unable to find a way to parse the value correctly if the record is at the beginning of the line. 6.8 is always assigned the value of 3.6.8. Is there any way to specify a preceding newline for those cases?

**Reply by gemu2015** [reply]
currently not.  i will think about a solution but may take a while

**Reply by gemu2015** [reply]
try this version, it should handle \r and \n escapes in obis decode

e.g.

1,\n6.8(@1,Zählerstand,MWh,mwh_count,3

[xsns_53_sml.ino.zip](https://github.com/user-attachments/files/26004447/xsns_53_sml.ino.zip)

**Reply by l33** [reply]
From the file you sent, I applied the following patch to 15.3.0. Using the `\n` prefix I was able to correctly parse the values I wanted.

```patch
--- a/tasmota/tasmota_xsns_sensor/xsns_53_sml.ino	2026-03-16 13:34:44
+++ b/tasmota/tasmota_xsns_sensor/xsns_53_sml.ino	2026-03-15 13:56:21
@@ -2121,7 +2135,20 @@
           continue;
         }
         if (sml_globs.mp[mindex].type == 'o' || sml_globs.mp[mindex].type == 'c') {
-          if (*mp++ != *cp++) {
+          // handle escapes
+          if (*mp == '/' && *(mp + 1) == 'n') {
+            // line feed escape
+            mp += 2;
+            if ('\n' != *cp++) {
+              found = 0;
+            }
+          } else if (*mp == '/' && *(mp + 1) == 'r') {
+            // carriage return escape
+            mp += 2;
+            if ('\r' != *cp++) {
+              found = 0;
+            }
+          } else if (*mp++ != *cp++) {
             found = 0;
           }
         } else {
```

Looking at the patch I read I would have to be using `/n` instead of `\n`. I am still puzzled why both actually work. There is probably a translation somewhere in the code but I was unable to find it.

In any case, @gemu2015 thank you! I hope this will be merged into a release at some point.
